crates/task-impls/src/view_sync.rs

```diff
@@ -344,14 +316,94 @@ where
+                }
+
+                let new_accumulator = VoteAccumulator2 {
+                    vote_outcomes: HashMap::new(),
+                    sig_lists: Vec::new(),
+                    signers: bitvec![0; self.exchange.total_nodes()],
+                    phantom: PhantomData,
                 };
 
-                if let Some(relay_task) = self.relay_task_map.get(&vote_internal.round) {
+                let mut relay_state = ViewSyncRelayTaskState::<
+                    TYPES,
+                    I,
+                    ViewSyncPreCommitVote<TYPES, ViewSyncMembership<TYPES, I>>,
+                    ViewSyncPreCommitCertificate2<TYPES>,
+                > {
+                    event_stream: self.event_stream.clone(),
+                    exchange: self.exchange.clone(),
+                    accumulator: either::Left(new_accumulator),
+                    id: self.id,
+                };
+
+                let result = relay_state.handle_event(event.clone()).await;
+
+                if result.0 == Some(HotShotTaskCompleted::ShutDown) {
+                    // The protocol has finished
+                    return;
+                }
+
+                relay_state = result.1;
+
+                let name = format!("View Sync Relay Task for view {:?}", vote.get_view_number());
+
+                let relay_handle_event = HandleEvent(Arc::new(
+                    move |event,
+                          state: ViewSyncRelayTaskState<
+                        TYPES,
+                        I,
+                        ViewSyncPreCommitVote<TYPES, ViewSyncMembership<TYPES, I>>,
+                        ViewSyncPreCommitCertificate2<TYPES>,
+                    >| {
+                        async move { state.handle_event(event).await }.boxed()
+                    },
+                ));
+
+                let filter = FilterEvent::default();
+                let builder = TaskBuilder::<
+                    ViewSyncRelayTaskStateTypes<
+                        TYPES,
+                        I,
+                        ViewSyncPreCommitVote<TYPES, ViewSyncMembership<TYPES, I>>,
+                        ViewSyncPreCommitCertificate2<TYPES>,
+                    >,
+                >::new(name)
+                .register_event_stream(relay_state.event_stream.clone(), filter)
+                .await
+                .register_registry(&mut self.registry.clone())
+                .await
+                .register_state(relay_state)
+                .register_event_handler(relay_handle_event);
+
+                let event_stream_id = builder.get_stream_id().unwrap();
+
+                self.relay_task_map
+                    .insert(vote.get_view_number(), ViewSyncTaskInfo { event_stream_id });
+                let _view_sync_relay_task = async_spawn(async move {
+                    ViewSyncRelayTaskStateTypes::build(builder).launch().await
+                });
+            }
+
+            HotShotEvent::ViewSyncCommitVoteRecv(vote) => {
```

Comment:
Same question as well here, are all the vote handler blocks the same but with different generic types for the relay task?

Reply:
Yes, they should all be the same.  I tried to make the `handle_event` function generic over the VOTE and CERT types to avoid this duplication, but the compiler complained when matching on events.  I think the solution is to break up the `handle_event` function so that each implementation of it only handles events related to 1 type of vote and certificate.  Definitely part of the clean up I need to do for view sync.